### ISSUE TITLE
Expected behavior of _upsampled_dft?

### ISSUE THREAD

**Issue by taylor-scott** [scikit-image/scikit-image#3735]
Continuing the discussion from [here](https://github.com/scikit-image/scikit-image/pull/3710#issuecomment-461640440). I'm happy to work on this, but I wanted to get feedback on exactly how this function is expected to behave before making any changes.

The behavior of _upsampled_dft is a little unexpected in my opinion and it's causing some overhead in register_translation.

#### Background
The function returns the result of: 
 - Embed the input image (already in Fourier space) in a larger image (also in Fourier space)
 - Take the DFT of the larger image
 - Return a small part of the DFT'ed image in an area of interest

Toy example:
```python
import numpy as np
from skimage.feature.register_translation import _upsampled_dft

image = np.arange(4*4).reshape((4, 4))  # Consider this to be in Fourier space
larger_image = np.zeros((8, 8))  # Also Fourier space, 0 frequencies centered
# Center the zero frequencies of image and embed in larger_image
larger_image[2:-2, 2:-2] = np.fft.fftshift(image)
# Return larger_image to 'standard order' for Fourier space
larger_image = np.fft.ifftshift(larger_image)

np.allclose(_upsampled_dft(image, 8, 2),  # Full output, upscale by 2
            np.fft.fftn(larger_image))  # True
```

To me at least, it's unexpected for the function to return the DFT of an image already in Fourier space. I would expect it either to take as input an image in real space or to return the inverse DFT of the image (which is what register_translation needs anyway).

#### My thoughts and questions:
 - Easy "fix": add an optional `inverse=False` flag
   - If True, return the inverse DFT of the image
   - This makes register_translation more explicit and faster for large images
   - If False, keep current behavior (which doesn't make sense for real inputs)
 - What's the most obvious or useful way to embed a real image in a larger image?
   - Centered in the larger image?
   - Upper left corner (i.e., `larger_image[:N, :M] = image`)
   - Is this functionality useful?
 - Making this function intuitive will probably require a breaking change
 - Either way, the docstring needs to be more explicit about how this function works

**Comment by sciunto** [comment]
xref: #3494

**Comment by hmaarrfk** [comment]
why are you considering the `image` and the `larger_image` to be in Fourier Space??? you are bringing them TO fourier space by calling `fftn` (notice the missing `i`).

**Comment by taylor-scott** [comment]
Yes, this is why I'm confused. And it's very possible this all makes perfect sense and I've just confused myself. :)

`_upsampled_dft` definitely takes the FFT of `image` embedded in a larger image.

The part I don't understand is how `image` is embedded in a larger image. As far as I can tell, it's a 3 step process:
1. Rearrange `image` with `np.fft.fftshift`
2. Place the rearranged `image` in the center of `larger_image`
3. Rearrange `larger_image` with `np.fft.ifftshift`

Then take the FFT of `larger_image`.

The only reason I can think of to do the `fftshift` and `ifftshift` parts is if `image` is already in Fourier space (which is what confuses me).

- Call `fftshift` to center the 0-frequency
- Place the image in the center of the larger image (so that the 0-frequency is still centered)
- Return the larger image to the standard order, with the 0-frequency at position 0.

Using the [numpy example for fftshift](https://docs.scipy.org/doc/numpy-1.15.0/reference/generated/numpy.fft.fftshift.html):

```python
>>> freqs = np.fft.fftfreq(10, 0.1)
>>> freqs
array([ 0.,  1.,  2.,  3.,  4., -5., -4., -3., -2., -1.])
>>> np.fft.fftshift(freqs)
array([-5., -4., -3., -2., -1.,  0.,  1.,  2.,  3.,  4.])
# Embed fftshift(freqs) in large_freqs (pad with 0)
>>>large_freqs
array([0, ..., 0, -5, -4, -3, -2, -1, 0, 1, 2, 3, 4, 0, ..., 0])
>>>np.fft.ifftshift(large_freqs)
array([0, 1, 2, 3, 4, 0, ..., 0, -5, -4, -3, -2, -1])
```

**Comment by hmaarrfk** [comment]
So let's try and use DFT for the operation and fft for the algorithm. 

Basically, upsampling a DFT is équivalent to zero padding then taking a Fourier transform.
That said, you have to zero pad in the right place. 
This depends on where `t=0 ` is.

There is also some funky things that go on because of the DFT assumes a periodic signal.

**Comment by hmaarrfk** [comment]
so the way you should think about it, is that when you provide a signal to the fft algorithm, it is really a signal centered around t=0, going from -T/2 to T/2 (integer rounding ignored).

**Comment by taylor-scott** [comment]
I agree that that is the expected behavior of an upsampled DFT: signal centered at `t=0`, going from `-T/2` to `T/2` with zero padding on either side.

I apologize, I'm struggling to explain the expected behavior. Toy examples of expected vs actual behavior (ignoring negative 0)

This is what I expect:
```python
a = np.arange(2*2).reshape(2, 2)
upsampled = _upsampled_dft(a, 4, 2)  # Full output, upsampled by 2
np.round(np.fft.ifftn(upsampled).real)
# array([[0, 0, 0, 0],
#        [0, 0, 1, 0],
#        [0, 2, 3, 0],
#        [0, 0, 0, 0]])
```

This is the actual output:
```python
a = np.arange(2*2).reshape(2, 2)
upsampled = _upsampled_dft(a, 4, 2)
np.round(np.fft.ifftn(upsampled).real)
# array([[0, 0, 0, 1],
#        [0, 0, 0, 0],
#        [0, 0, 0, 0],
#        [2, 0, 0, 3]])
```

The resulting matrix can be recreated by:
```python
a = np.arange(2*2).reshape(2, 2)
embedded = np.zeros((4, 4))
embedded[1:-1, 1:-1] = np.fft.fftshift(a)
embedded = np.fft.ifftshift(embedded)
embedded
# array([[0., 0., 0., 1.],
#        [0., 0., 0., 0.],
#        [0., 0., 0., 0.],
#        [2., 0., 0., 3.]])

```

**Comment by hmaarrfk** [comment]
Gah, sorry, this is likely due to a different definition of where the time samples come from:

```
[0dt, 1dt, 2dt, 3dt]
```
vs
```
[0dt, 1dt, -2dt, -1dt]
```
both are equivalent, but different definitions require fftshifts to get working correctly. I think matlab likes to use the second version.

```
In [7]: upsampled_np = np.fft.fftn(a, (4, 4))                                                                                                                        

In [8]: np.round(np.fft.ifftn(upsampled_np).real)                                                                                                                    
Out[8]: 
array([[0., 1., 0., 0.],
       [2., 3., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.]])
```

**Comment by taylor-scott** [comment]
So is that last example what we expect of `_upsampled_dft`? This is also how Matlab does it:
```matlab
Y = reshape(1:4, [2, 2]);
real(ifftn(fftn(Y, [4, 4])))
ans = 4x4
       1    2    0    0
       3    4    0    0
       0    0    0    0
       0    0    0    0
```

**Comment by hmaarrfk** [comment]
i'm not too sure in this case. it depends what you want. I would look at how they use the **results** of the upsampled dft.

**Comment by taylor-scott** [comment]
Confusingly, this function is actually used to calculate the *inverse* upsampled DFT in `register_translation`...
```python
cross_correlation = _upsampled_dft(image.conj(), ...).conj()
```

The embedding makes sense here but it's a roundabout way to calculate the inverse DFT. The `conj()` calls can be avoided if `_upsampled_dft` (optionally) returns the inverse DFT.

E.g.
- If `inverse=True`
    - Keep embedding as is
    - `kernel = np.exp(im2pi * kernel)` to give inverse DFT
- If `inverse=False`
    - More obvious embedding?
    - `kernel = np.exp(-im2pi * kernel)` as it is now

**Comment by hmaarrfk** [comment]
This is likely an ugly hack, because the people working with this likely only looked at the magnitude of the cross correlation.


https://github.com/scikit-image/scikit-image/blob/master/skimage/feature/register_translation.py#L207

**Comment by hmaarrfk** [comment]
i noticed a few other funny things in this code when going through it myself:

Lets simplify the problem to the 1D case.

I learned that there are 2 possible solutions for the interpretations to what the index of the maximum of the fourier transform gives you. Lets say we have a length N signal, and the maximum is at index m.

1. A shift of `m` in the `+` direction.
2. A shift of `N-m` in the `-` direction

If `m` is greater than `N/2` and we assume "small shifts", then Option 2 is probably the correct one.

Typically, people use a "real space" approach as a final check. I think this was explained in the implementation details of the MIST document. 

here, this check is omitted.

It might be a clue as to why the original author made such decisions.

**Comment by jni** [comment]
@hmaarrfk fascinating! Are you saying we don't "simplify" to Option 2 currently? That would be bad.

**Comment by hmaarrfk** [comment]
No, i think the correct code takes a shortcut to guessing option 2 (maybe). I'm not too sure. My code uses Scikit-image's algorithms to some extent and have found some oddities from time to time. Not enough to make me file a bug request though.

**Comment by taylor-scott** [comment]
@hmaarrfk Can you explain the real space check part? Or link to the MIST document? I'm not sure what that's referring to.

I think we implement selection Option 1 vs 2 with [this line](https://github.com/scikit-image/scikit-image/blob/7f8d37d9ace5effe318b1808c32970fcb608ddf3/skimage/feature/register_translation.py#L185)

```python
    shifts = np.array(maxima, dtype=np.float64)
    shifts[shifts > midpoints] -= np.array(shape)[shifts > midpoints]
```

**Comment by hmaarrfk** [comment]
![image](https://user-images.githubusercontent.com/90008/52920932-ddb17080-32df-11e9-8b23-75bee9622ce1.png)
see the "technical documentation link"
https://github.com/usnistgov/MIST

**Comment by hmaarrfk** [comment]
@taylor-scott, correct. but I think we should technically be measure the autocorrelation in the real space instead.

**Comment by taylor-scott** [comment]
I think we do measure correlation in real space?

 For the pixel registration:
```python
cross_correlation = np.fft.ifftn(image_product)
```

And in subpixel registration:
```python
cross_correlation = _upsampled_dft(image_product.conj(), ...).conj() / normalization
```
([equivalent](https://en.wikipedia.org/wiki/Discrete_Fourier_transform#Expressing_the_inverse_DFT_in_terms_of_the_DFT) to `_upsampled_inverse_dft(image_product, ...)` )

**Comment by taylor-scott** [comment]
Or are you talking about the Interpret-Translation algorithm? I don't think we implement that

**Comment by hmaarrfk** [comment]
I'm not a big fan of algorithm names, usually when something is obvious, it is obvious.

The idea is that for discrete signals, frequencies are wrapped around 2pi. So +x, is the same as (-2pi + x). We need to compare, for each axis, the result of the 2 possible translations.

![image](https://user-images.githubusercontent.com/90008/52922574-13f6ec00-32f0-11e9-9cba-1e78b82de969.png)

for the 2D case, we need to compare 4 translations, as they show in the algorithm above.

In either case, I think this derails from your oginal request, that "upsample_dft" return the result in the "non-fftshifted" element order, which is probably reasonable.

As for conjugate or non-conjugate, maybe that is better addressed via comments???

**Comment by taylor-scott** [comment]
I agree, I think this is verging on another discussion. 

Thanks for the explanation about the different definitions of time samples. 

I'll open a PR to add the inverse flag (and clear up the `conj()` hack) but leave the non-inverse as is.
